raft/Raft.tla

```diff
@@ -0,0 +1,247 @@
+    /\ UNCHANGED <<voters, hardState, commitHistory>>
+
+(* Checks if it's possible to commit logs. It's the same as handling MsgAppendResponse. *)
+Commit(v) ==
+    /\ LastLogOf(v).index # hardState[v].commit
+    /\ LET quorumLogs == { l \in logs[v]: /\ l.index > hardState[v].commit
+                                          /\ LET met == { o \in voters: l \in logs[o] }
+                                             IN Cardinality(met) * 2 > Cardinality(voters)}
+       IN /\ quorumLogs # {}
+          /\ LET quorumLastLog == LastLog(quorumLogs)
+             IN /\ quorumLastLog.term = hardState[v].term
+                /\ hardState' = [hardState EXCEPT ![v] = [@ EXCEPT !.commit = quorumLastLog.index]]
+                /\ commitHistory' = commitHistory \union { l \in logs[v]: /\ l.index <= quorumLastLog.index
+                                                                          /\ l.index > hardState[v].commit }
+                /\ UNCHANGED <<voters, softState, logs, leaderHistory>>
+
+StepFollower(v) == \/ CheckLeader(v)
+                   \/ FetchLog(v)
+                   \/ Vote(v)
+
+StepCandidate(v) == \/ CheckLeader(v)
+                    \/ Vote(v)
+                    \/ ClaimLeadership(v)
+
+StepLeader(v) == \/ CheckLeader(v)
+                 \/ Commit(v)
+                 \/ Vote(v)
+
+Step(v) == CASE softState[v].role = "Leader" -> StepLeader(v)
+           [] softState[v].role = "Candidate" -> StepCandidate(v)
+           [] softState[v].role = "Follower" -> StepFollower(v)
+
+-----------------------------------------------------------------------
+
+Next == \E v \in voters: \/ Step(v)
+                         \/ Tick(v)
+
+Spec == Init /\ [][Next]_vars
+
+-----------------------------------------------------------------------
+
+(* Election Safety: at most one leader can be elected in a given term. *)
+ElectionSafety ==
+    \A h1, h2 \in leaderHistory: \/ h1.term # h2.term
+                                 \/ h1.leader = h2.leader
+
+CommittedLog(v) ==
+    { l \in logs[v]: l.index <= hardState[v].commit }
+
+(* State Machine Safety: if a server has applied a log entry  *)
+(* at a given index to its state machine, no other server     *)
+(* will ever apply a different log entry for the same index.  *)
+StateMachineSafety ==
+    (* Log Matching: if two logs contain an entry with the same    *)
+    (* index and term, then the logs are identical in all entries  *)
+    (* up through the given index.                                 *)
+    /\ \A l1, l2 \in commitHistory: \/ l1.index # l2.index
+                                    \/ l1 = l2
+    (* Leader Completeness: if a log entry is committed in a given *)
+    (* term, then that entry will be present in the logs of the    *)
+    (* leaders for all higher-numbered terms.                      *)
+    /\ \A v \in voters: \/ softState[v].role # "Leader"
+                        \/ CommittedLog(v) = { l \in commitHistory: l.index <= hardState[v].commit }
```

Comment:
Does neither the role is leader or others, as long as `CommittedLog(v) = { l \in commitHistory: l.index <= hardState[v].commit }` is true then true. seems this is not to say `Leader Completeness`?

Reply:
It's the same as
```
if isLeader then
   it contains all committed logs in the history
else
   true
fi
```

Though it may be changed to if to reduce calculation.

Reply:
Yes, I think it's better to change to `if` to reduce calculation and maybe more easy to understand, such as:
```
\/  /\ softState[v].role = 'Leader'
    /\ CommittedLog(v) = { l \in commitHistory: l.index <= hardState[v].commit }
\/  /\ softState[v].role #1 'Leader'
```

Reply:
Updated. Seem no improvement is observed.

Reply:
:(